grails-app/services/org/pih/warehouse/core/IdentifierService.groovy

```diff
@@ -150,13 +153,83 @@ class IdentifierService {
         return identifier.toUpperCase()
     }
 
+    def generateSequentialProductIdentifier(ProductType productType, String sequentialPatternChar, Integer allowedSequences) {
+        // If custom identifier contains more than one sequential part, then throw exception
+        Pattern pattern = Pattern.compile("${sequentialPatternChar}+")
+        Matcher matcher = pattern.matcher(productType.productIdentifierFormat)
+        int count = 0
+        def sequenceFormat = ""
+        while (matcher.find()) {
+            sequenceFormat = matcher.group()
+            count++
+        }
+        if (count > allowedSequences) {
+            throw new IllegalArgumentException("Cannot have more sequence numbers than ${allowedSequences} in the same identifier")
+        }
+
+        def sequenceNumber = productType.getNextSequenceNumber()
```

Comment:
At what point do we save this to the database? And how do we ensure that two threads don't try to save the same identifier. Again, having this fail on a write operation (i.e. unique constraint on product.productCode) rather than on the read operation is fine with me for now but it might be good to have another solution in mind for maybe locking on read at some point. I don't know how other systems handle this so maybe we can create a spike to look into it.